File: python-backend/app/repositories/supplier_repository.py
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models import Supplier


class SupplierRepository:
    def __init__(self, session: AsyncSession):
        self.session = session

    async def list_suppliers(self, *, organization_id: str = "org_1", active_only: bool = True) -> list[Supplier]:
        query = select(Supplier).where(Supplier.organization_id == organization_id)
        if active_only:
            query = query.where(Supplier.is_active.is_(True))
        query = query.order_by(Supplier.name.asc())
        result = await self.session.execute(query)
        return list(result.scalars().all())

    async def get_supplier(self, supplier_id: str) -> Supplier | None:
        return await self.session.get(Supplier, supplier_id)

    async def update_supplier(
        self,
        supplier: Supplier,
        *,
        name: str,
        website_url: str | None,
        integration_type: str,
        contact_name: str | None,
        contact_email: str | None,
    ) -> Supplier:
        supplier.name = name
        supplier.website_url = website_url
        supplier.integration_type = integration_type
        supplier.contact_name = contact_name
        supplier.contact_email = contact_email
        await self.session.commit()
        await self.session.refresh(supplier)
        return supplier

File: server/repositories/supplierRepository.js
const path = require("path");
const { DatabaseSync } = require("node:sqlite");

const databasePath = path.resolve(__dirname, "../../prisma/dev.db");

function openDatabase() {
    const database = new DatabaseSync(databasePath);
    database.exec("PRAGMA foreign_keys = ON; PRAGMA busy_timeout = 5000; PRAGMA journal_mode = WAL;");
    return database;
}

function mapSupplierRow(row) {
    if (!row) {
        return null;
    }

    return {
        id: row.id,
        organization_id: row.organization_id,
        name: row.name,
        code: row.code,
        website_url: row.website_url,
        integration_type: row.integration_type,
        is_active: Boolean(row.is_active),
        contact_name: row.contact_name,
        contact_email: row.contact_email,
        created_at: row.created_at,
        updated_at: row.updated_at
    };
}

function listSuppliers({ organizationId = "org_1", activeOnly = true } = {}) {
    const database = openDatabase();

    try {
        const conditions = ["organization_id = ?"];
        const params = [organizationId];

        if (activeOnly) {
            conditions.push("is_active = 1");
        }

        const rows = database.prepare(`
            SELECT *
            FROM suppliers
            WHERE ${conditions.join(" AND ")}
            ORDER BY name ASC
        `).all(...params);

        return rows.map(mapSupplierRow);
    } finally {
        database.close();
    }
}

function getSupplierById(supplierId) {
    const database = openDatabase();

    try {
        const row = database.prepare(`
            SELECT *
            FROM suppliers
            WHERE id = ?
        `).get(supplierId);

        return mapSupplierRow(row);
    } finally {
        database.close();
    }
}

function updateSupplier(supplierId, changes) {
    const database = openDatabase();

    try {
        const existing = database.prepare(`
            SELECT *
            FROM suppliers
            WHERE id = ?
        `).get(supplierId);

        if (!existing) {
            return null;
        }

        const timestamp = new Date().toISOString();
        const nextValues = {
            name: Object.prototype.hasOwnProperty.call(changes, "name")
                ? changes.name
                : existing.name,
            website_url: Object.prototype.hasOwnProperty.call(changes, "website_url")
                ? changes.website_url
                : existing.website_url,
            integration_type: Object.prototype.hasOwnProperty.call(changes, "integration_type")
                ? changes.integration_type
                : existing.integration_type,
            contact_name: Object.prototype.hasOwnProperty.call(changes, "contact_name")
                ? changes.contact_name
                : existing.contact_name,
            contact_email: Object.prototype.hasOwnProperty.call(changes, "contact_email")
                ? changes.contact_email
                : existing.contact_email
        };

        database.prepare(`
            UPDATE suppliers
            SET
                name = ?,
                website_url = ?,
                integration_type = ?,
                contact_name = ?,
                contact_email = ?,
                updated_at = ?
            WHERE id = ?
        `).run(
            nextValues.name,
            nextValues.website_url,
            nextValues.integration_type,
            nextValues.contact_name,
            nextValues.contact_email,
            timestamp,
            supplierId
        );

        return getSupplierById(supplierId);
    } finally {
        database.close();
    }
}

module.exports = {
    listSuppliers,
    getSupplierById,
    updateSupplier
};
